File: edge/src/ratelimit.js
// Distributed rate limiting in Workers KV. ARCHITECTURE.md §11.3.
//
// Three composite buckets, all checked:
//
//   ip:<prefix>   /24 for IPv4, /48 for IPv6 — NOT a single address.
//                 Carrier-grade NAT in India means an entire operator region can
//                 share addresses, and a busy pharmacy's Wi-Fi is one IP. Keying
//                 on a single address throttles real consumers before it
//                 throttles attackers (F21).
//
//   tag:<hash>    hash of the m parameter's UID portion. A single pack verified
//                 500 times in an hour from 500 addresses is the signal you
//                 actually want, and a per-IP limiter cannot see it (G1, D21).
//
//   sess:<cookie> soft throttle for real users — slow down, don't block.
//
// KV is eventually consistent, so these limits are approximate. That is fine: it
// is still far stronger than v1's per-process in-memory counters, which were
// silently N times looser than configured (F16). The Postgres limiter in
// backend/ratelimit.py is the second layer for direct-to-origin traffic.

export const LIMITS = {
  ip: { window: 60, max: 60 },
  tag: { window: 3600, max: 30 },
  sess: { window: 60, max: 30 },
};

export function ipPrefix(request) {
  const ip = request.headers.get('CF-Connecting-IP') || '';
  if (!ip) return 'unknown';
  if (ip.includes(':')) {
    // IPv6 -> /48, the first three hextets.
    return ip.split(':').slice(0, 3).join(':') + '::/48';
  }
  return ip.split('.').slice(0, 3).join('.') + '.0/24';
}

export async function tagBucket(m) {
  // Hash the UID rather than keying on it directly, so the KV namespace itself
  // is not a list of every UID that has ever been scanned.
  const data = new TextEncoder().encode('nfcmed/v2/edge-tag:' + m.slice(0, 14));
  const digest = await crypto.subtle.digest('SHA-256', data);
  return [...new Uint8Array(digest).slice(0, 12)]
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('');
}

function windowStart(seconds) {
  return Math.floor(Date.now() / 1000 / seconds) * seconds;
}

/**
 * Fixed-window counter in KV. Returns { ok, retryAfter, bucket }.
 * Fails OPEN when KV is unavailable — the origin's Postgres limiter is the
 * backstop, and a KV outage must not take the consumer path down. An unreachable
 * verification service IS a security failure (F22), not just an inconvenience.
 */
export async function checkBucket(kv, kind, identity) {
  const { window, max } = LIMITS[kind];
  const start = windowStart(window);
  const key = `rl:${kind}:${identity}:${start}`;
  try {
    const current = parseInt((await kv.get(key)) || '0', 10);
    if (current >= max) {
      const retryAfter = start + window - Math.floor(Date.now() / 1000);
      return { ok: false, retryAfter: Math.max(retryAfter, 1), bucket: kind };
    }
    // expirationTtl cleans up on its own, so the namespace never needs sweeping.
    await kv.put(key, String(current + 1), { expirationTtl: window + 60 });
    return { ok: true };
  } catch (err) {
    return { ok: true, degraded: true };
  }
}

export async function checkAll(kv, request, m) {
  const checks = [
    checkBucket(kv, 'ip', ipPrefix(request)),
    checkBucket(kv, 'tag', await tagBucket(m)),
  ];
  for (const result of await Promise.all(checks)) {
    if (!result.ok) return result;
  }
  return { ok: true };
}

File: backend/ratelimit.py
"""Postgres-backed second-layer rate limiter. ARCHITECTURE.md §9, §11.3.

The FIRST layer is the Cloudflare Worker's KV limiter, which is global and sees
per-tag traffic across every edge location. This layer exists for the case where
the edge is bypassed — someone who discovers the Render origin and hits it
directly.

Why not flask-limiter's memory:// store, as v1 used? Because with gunicorn
--workers 2 each worker keeps its own counter, so the configured limit was
silently 2x looser than reported (F16). Any evaluation that reports rate-limiter
behaviour has to state the worker count, or the number means nothing. A shared
store removes the caveat entirely, and Postgres is the shared store we already
pay for.

Fixed windows, not sliding: a fixed window can let through up to 2x the limit
across a boundary. That is an acceptable, documented imprecision at this layer —
this is a backstop, not the primary control.

IP KEYING IS BY PREFIX (/24 IPv4, /48 IPv6), NOT BY ADDRESS. Carrier-grade NAT
in India means an entire operator region can share addresses and a busy
pharmacy's Wi-Fi is one IP. Keying on a single address throttles real consumers
before it throttles attackers (F21).
"""
from __future__ import annotations

import ipaddress
import logging
from datetime import datetime, timedelta, timezone

import db
import metrics
from errors import RateLimited

log = logging.getLogger(__name__)

# Origin-side limits from the §9.4 route inventory.
LIMITS = {
    "verify": (120, 60),   # 120 requests per 60 s
    "enrol": (60, 60),
    "reenrol": (5, 60),
    "report": (5, 60),
    "admin": (60, 60),
    "metrics": (60, 60),
}

_UPSERT = """
INSERT INTO rate_limit_bucket (bucket_key, window_start, hits)
VALUES (%s, %s, 1)
ON CONFLICT (bucket_key, window_start)
DO UPDATE SET hits = rate_limit_bucket.hits + 1
RETURNING hits
"""


def ip_prefix(ip: str | None) -> str:
    """/24 for IPv4, /48 for IPv6. Returns 'unknown' for anything unparseable —
    which then shares one bucket, deliberately: an unparseable source is more
    likely to be hostile than to be a pharmacy."""
    if not ip:
        return "unknown"
    candidate = ip.split(",")[0].strip()
    try:
        addr = ipaddress.ip_address(candidate)
    except ValueError:
        return "unknown"
    if addr.version == 4:
        return str(ipaddress.ip_network(f"{addr}/24", strict=False))
    return str(ipaddress.ip_network(f"{addr}/48", strict=False))


def _window_start(window_seconds: int, now: datetime | None = None) -> datetime:
    now = now or datetime.now(timezone.utc)
    epoch = int(now.timestamp()) // window_seconds * window_seconds
    return datetime.fromtimestamp(epoch, tz=timezone.utc)


def check(scope: str, identity: str) -> None:
    """Raise RateLimited (429 + Retry-After) if this identity is over the limit.

    Fails OPEN on a database error. That is a deliberate trade: this is the
    second of two layers, and an unreachable database already means the request
    is about to fail anyway. Failing closed here would turn a database blip into
    a total outage of the consumer path — and the consumer path being down IS a
    security failure (F22), not just an inconvenience.
    """
    limit, window = LIMITS.get(scope, (60, 60))
    bucket_key = f"{scope}:{identity}"
    start = _window_start(window)
    try:
        with db.connection(write=True) as conn, conn.cursor() as cur:
            cur.execute(_UPSERT, (bucket_key, start))
            hits = cur.fetchone()[0]
    except Exception as exc:
        log.warning("ratelimit_unavailable",
                    extra={"error": exc.__class__.__name__, "scope": scope})
        return

    if hits > limit:
        metrics.incr("rate_limited_total")
        retry_after = int((start + timedelta(seconds=window)
                           - datetime.now(timezone.utc)).total_seconds()) + 1
        raise RateLimited(f"{scope} limit {limit}/{window}s exceeded",
                          retry_after=max(retry_after, 1))


def sweep(older_than_seconds: int = 3600) -> int:
    """Delete expired windows. Called from the nightly Action so the table does
    not grow without bound on a free-tier database."""
    cutoff = datetime.now(timezone.utc) - timedelta(seconds=older_than_seconds)
    try:
        with db.connection(write=True) as conn, conn.cursor() as cur:
            cur.execute("DELETE FROM rate_limit_bucket WHERE window_start < %s", (cutoff,))
            return cur.rowcount
    except Exception as exc:
        log.warning("ratelimit_sweep_failed", extra={"error": exc.__class__.__name__})
        return 0
